Add basic SSH documentation for CLI (#15316)
Lightning-AI/lightning: docs/source-app/workflows/ssh/index.rst

#################################
Debug cloud apps via SSH
#################################

**Audience:** Users that want to debug their cloud apps on their own machine

> You can

Reviewer: this needs more context

Author: tried to give more context - LMK if you have ideas how to phrase this better.

   $ eval "$(ssh-agent -s)"
   > Agent pid 12345

Add your generated ssh key:

.. code:: bash

   $ ssh-add ~/.ssh/id_ed25519

Lastly, verify your ssh-key is properly loaded:

.. code:: bash

   $ ssh-add -L
   ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIAn8mYRnRG1banQcfXPCUC6R8FvQS+YgfIsl70/dD3Te your_email@example.com

Step 3: Add your key to Lightning
=================================

You can add SSH keys using Lightning.ai UI, or via this CLI command:

.. code:: bash

   $ lightning add ssh-key --public-key ~/.ssh/id_ed25519.pub

Next, list your configured SSH keys:

.. code:: bash

   $ lightning list ssh-keys

You are now ready to access your Lightning Flow and Work containers.

----

**********************************************************
SSH to your cloud app
**********************************************************

Ensure you have a running Lightning application in the cloud:

.. code:: bash

   $ lightning run app app.py --cloud --name my-app

You can SSH any app you are running on the cloud.

To view all apps you can simple use this following:

.. code:: bash

   $ lightning ssh

.. code:: bash

   $ lightning ssh

To connect to an app flow use:

.. code:: bash

>    $ lightning ssh --app-id <your-app-id>

Reviewer: where can users find their app-id?

Author: added context explaining how to get the app id

To connect to a LightningWork component use:

.. code:: bash

>    $ lightning ssh --app-id <your-app-id> --component-name flow

Reviewer: where can user find component name?

Author: added an explanation how to find the component name.

Reviewer 2: You instruct to connect to a work but have `--component-name flow` in the example

Author: @awaelchli  you can SSH into both work and flow . `flow` is always a component available to ssh into, namely your root flow container.

> #################################

Author: done :+1:

Debug cloud apps via SSH
#################################

**Audience:** Users that want to debug their cloud apps on their own machine.

SSH allows you to access all your Lightning components in real time.
This can be useful if you execute your app in a different environment than you develop it.
This is often the case with applications that rely on GPUs.

You can inspect the entire file system and monitor the runtime environment.

----

**********************************************************
Add SSH key to lightning
**********************************************************

> Before you can SSH to cloud machines, you will need to generate a new private SSH key, add it to the SSH agent, and add the public SSH key to your account on Lightning.

Reviewer 2: Maybe mention here that the following three steps are one-time only? Or do you think it is clear to the user? Maybe it is

Author: Depends on the reader - I've added a sentence here in case an unexperienced user follows this to.

Step 1: Create an SSH key
==========================

Open a terminal and run the following command (replace email with the address you used in your lightning.ai account):

.. code:: bash

   # make the ssh key (if you don't have one)
   $ ssh-keygen -t ed25519 -C "your_email@example.com"

This creates a new SSH key, using the provided email as a label.

At the prompt, type a secure passphrase.

.. code:: bash

   > Enter passphrase (empty for no passphrase): [Type a passphrase]
   > Enter same passphrase again: [Type passphrase again]

> Step 2: add the key to the ssh-agent

Reviewer 2: I find it strange that this step is in this section. This step is not a one-time setup and needs to be repeated every time you re-login to your system. I would expect it to be mentioned in the section where we ssh into the machine.

Author: I was skeptical about even adding this in the first place, as it is setup dependent.
Ie, with a hardware key like a yubikey, you don't need to do anything.

I'll move it around, as that is an uncommon setup I think.

Step 3: Add your key to Lightning
=================================

You can add SSH keys using Lightning.ai UI, or via this CLI command:

.. code:: bash

   $ lightning add ssh-key --public-key ~/.ssh/id_ed25519.pub

Next, list your configured SSH keys:

.. code:: bash

   $ lightning list ssh-keys

You are now ready to access your Lightning Flow and Work containers.

----

**********************************************************
SSH to your cloud app
**********************************************************

Ensure you have a running Lightning application in the cloud:

.. code:: bash

   $ lightning run app app.py --cloud --name my-app

You can SSH any app you are running on the cloud.

To view all apps you can simple use this following:

.. code:: bash

   $ lightning ssh

.. code:: bash

   $ lightning ssh

To connect to an app flow use:

.. code:: bash

   $ lightning list apps
   $ lightning ssh --app-id <your-app-id> # taken from previous app listing

To connect to a LightningWork component use:

.. code:: bash

   $ lightning ssh --app-id <your-app-id> --component-name flow

> The component name is the variable name of your LightningWork instances in python.

Reviewer 2: ```suggestion
The component name is the variable name of your LightningWork instances in Python.
```

You can inspect the entire file system and monitor the runtime environment.

----

************************
Add SSH key to Lightning
************************

Before you can SSH to cloud machines, you will need to generate a new private SSH key, add it to the SSH agent, and add the public SSH key to your account on Lightning.
The following steps are one-time setup to allow you to SSH into your cloud machines.

Step 1: Create an SSH key
==========================

Open a terminal and run the following command (replace email with the address you used in your lightning.ai account):

.. code:: bash

   # make the ssh key (if you don't have one)
   $ ssh-keygen -t ed25519 -C "your_email@example.com"

This creates a new SSH key, using the provided email as a label.

At the prompt, type a secure passphrase.

.. code:: bash

   > Enter passphrase (empty for no passphrase): [Type a passphrase]
   > Enter same passphrase again: [Type passphrase again]

Step 2: Add your key to Lightning
=================================

You can add SSH keys using Lightning.ai UI, or via this CLI command:

.. code:: bash

   $ lightning add ssh-key --public-key ~/.ssh/id_ed25519.pub

> Next, list your configured SSH keys:

Reviewer: why this step?
if keeping it, explain how to validate

Author: this step is done to validate the SSH key was added properly. I'm fine dropping it.

#################################

> Access your Apps via SSH

Reviewer: ```suggestion
Debug cloud apps via SSH 
```

#################################

> **Audience:** Users who want to debug their Flow and Work in a the runtime environment.

Reviewer: ```suggestion
**Audience:** Users that want to debug their cloud apps on their own machine
```

----

> 

Reviewer: Add motivation-

When running your app on the cloud, it could be useful to SSH to cloud machines so you can debug them on your local machine as if they are running locally:
• Inspect files
• Inspect environment
• Monitor resources
etc

**********************************************************

> Generating a new SSH key and configure ssh-agent

Reviewer: Titles should be verbs, see style guide: https://www.notion.so/lightningai/Check-page-structure-4c4e660c70b949a7aa75763e6d0955e9

> Generating a new SSH key and configure ssh-agent

Reviewer: ```suggestion
Add SSH key to lightning
```

**********************************************************

> Open a terminal and run the following command:
> 
> .. code:: bash
> 
>    $ ssh-keygen -t ed25519 -C "your_email@example.com"
> 
> This creates a new SSH key, using the provided email as a label.
> 
> At the prompt, type a secure passphrase.
> 
> .. code:: bash
> 
>    > Enter passphrase (empty for no passphrase): [Type a passphrase]
>    > Enter same passphrase again: [Type passphrase again]
> 
> Next, start your ssh-agent in the background:
> 
> .. code:: bash
> 
>    $ eval "$(ssh-agent -s)"
>    > Agent pid 12345
> 
> Add your generated ssh key:
> 
> .. code:: bash
> 
>    $ ssh-add ~/.ssh/id_ed25519
> 
> Lastly, verify your ssh-key is properly loaded:
> 
> .. code:: bash
> 
>    $ ssh-add -L
>    ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIAn8mYRnRG1banQcfXPCUC6R8FvQS+YgfIsl70/dD3Te your_email@example.com
> 
> ----
> 
> **********************************************************
> Managing SSH keys with lightning CLI
> **********************************************************
> Open a terminal and run the following command:
> 
> .. code:: bash
> 
>    $ lightning add ssh-key --public-key ~/.ssh/id_ed25519.pub
> 
> Next, list your configured SSH keys:
> 
> .. code:: bash
> 
>    $ lightning list ssh-keys
> 
> You are now ready to access your Lightning Flow and Work containers.
> 
> ----

Reviewer: ```suggestion

Before you can SSH to cloud machines, you will need to generate a new private SSH key, add it to the SSH agent, and add the public SSH key to your account on Lightning.

Step 1: Create an SSH key
===================

Open a terminal and run the following command (replace email with the address you used in your lightning.ai account):

.. code:: bash

   # make the ssh key (if you don't have one)
   $ ssh-keygen -t ed25519 -C "your_email@example.com"

This creates a new SSH key, using the provided email as a label.

At the prompt, type a secure passphrase.

.. code:: bash

   > Enter passphrase (empty for no passphrase): [Type a passphrase]
   > Enter same passphrase again: [Type passphrase again]

Step 2: add the key to the ssh-agent
===========================

Next, start your ssh-agent in the background:

.. code:: bash

   # add the key to the ssh-agent (to avoid having to explicitly state key on each connection)
   # to start the agent, run the following
   $ eval "$(ssh-agent -s)"
   > Agent pid 12345

Add your generated ssh key:

.. code:: bash

   $ ssh-add ~/.ssh/id_ed25519

Lastly, verify your ssh-key is properly loaded:

.. code:: bash

   $ ssh-add -L
   ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIAn8mYRnRG1banQcfXPCUC6R8FvQS+YgfIsl70/dD3Te your_email@example.com

Step 3: Add your key to Lightning
=========================

You can add SSH keys using Lightning.ai UI, or via this CLI command:

.. code:: bash

   $ lightning add ssh-key --public-key ~/.ssh/id_ed25519.pub

Next, list your configured SSH keys:

.. code:: bash

   $ lightning list ssh-keys

You are now ready to access your Lightning Flow and Work containers.

----
```

**********************************************************

> Accessing your Lightning Flow and Work containers via SSH

Reviewer: ```suggestion
SSH to your cloud app
```

**********************************************************

Ensure you have a running Lightning application in the cloud:

.. code:: bash

   $ lightning run app app.py --cloud --name my-app

Now, use the CLI to access any of the available components:

.. code:: bash

>    $ lightning ssh

Reviewer: You can SSH any app you are running on the cloud.

To view all apps you can simple use this following:

.. code:: bash
   $ lightning ssh

To connect to an app flow use:

.. code:: bash
   $ lightning ssh --app-id bla

To connect to a LightningWork component use:

.. code:: bash
   $ lightning ssh --app-id bla --component-name bla
